### ISSUE TITLE
Replacement of mgv7 floatlands

### ISSUE THREAD

**Issue by paramat** [minetest/minetest#9076]
I intend to remove the 'highly unstable' floatlands from mgv7, this issue explains why.
I will leave this issue up for a few weeks before starting work, if core devs have objections please discuss before i start work.

The mgv7 floatlands were added, to my surprise, as long as 3 years ago bc3980e1d69ca9f848f80eb7386525fa4fb326cd
They were added in an unfinished and untuned state and have always been disabled by default and described as 'highly unstable'.

PR #9026 was intended to address the mistake of reusing mgv7 surface mountain noise for floatland mountains.
This reuse was 'clever' in a sense but problematic:
The ideal noise parameters for one are far from ideal for the other.
Any customisation of noise for one feature will completely mess up the other feature.
I also became tired of those crazy mgv7 mountain 3D noise shapes (as many others are) and did not want those repeated for the floatlands.

But PR 9026 has a big problem, the only rewrite i found acceptable created a very dangerous mess in existing worlds if the mapgen parameters in map_meta.txt are not altered on changeover (which will certainly happen on some servers). Vast volumes of water start to fall towards the world surface, enough to harm server performance and create near-unfixable water griefing on the surface.

In that PR i began to realise more problems and mistakes:
* Mgv7 is a very old mapgen, and is now the default mapgen, so i consider it a mistake to continue to mess with it instead of moving on to newer mapgens.
* The whole concept of adding floatlands is to make use of all the wasted space above the world surface. These floatlands are only a single level at y = 1280, they should be present throughout the atmosphere.
* Adding water to floatlands is, i now realise, extremely dangerous. I now feel that random  floatlands placed above the surface should probably be waterless. If water is going to be present on multiple levels i need to integrate that very carefully in a new mapgen designed for that.
* Adding another lake level to the mapgen requires complex biome code in games:
.......
A complete set of floatland lake and land biomes have to be defined.
The standard set of biomes have to be altered to stop just below the floatlands.
The floatland biomes now extend from the floatland all the way up to y = 31000, but this may be a problem. If this problem is solved we would need to register yet another complete set of standard land biomes to have effect higher up.
Any decorations and ores that are limited to certain biomes now need the floatland biomes included in their biome list.
.......
Now consider that if floatlands were waterless, biome code would not need any alteration, floatlands would just be treated like the tops of mountains or the floating rocks found in mgv7. All that mess in MTG and games can be removed, making game creation much easier for everyone.
All the extra biome code in MTG is a pain, and lot just to support 1 mapgen with this type of floatland.
* The complex biome code means that most games, which do not have it, are incompatible with mgv7 floatlands.

I suggest that mgv7 floatlands are removed, with plenty of warning on the forum, so that anyone using them is prepared to build walls across any half-generated lakes (this should be fairly rare) to avoid leaks.

I am still, of course, committed to mapgen features that make full use of all the space in a world, but will do this in newer mapgens or future mapgens.

**Comment by FaceDeer** [comment]
That's fair, I suppose.

Are there any features of the Floatlands that would be difficult to extract into a Lua mapgen mod? Having the Floatlands as a separate mod might make it much more versatile and amenable to tweaking.

**Comment by sfan5** [comment]
I don't think floatlands absolutely need to be removed from mgv7, some of these problems are fixable.
Perhaps further development should happen in a newer mapgen, while keeping old code around.

> I also became tired of those crazy mgv7 mountain 3D noise shapes (as many others are) and did not want those repeated for the floatlands.

This is subjective, I think floatlands as they are now look okay even if they use the same "crazy" mountain noise.
Most (even advanced) players will have never touched floatlands, so maybe some more feedback would be nice?

> But PR 9026 has a big problem, the only rewrite i found acceptable created a very dangerous mess in existing worlds if the mapgen parameters in map_meta.txt are not altered on changeover (which will certainly happen on some servers).

Can't the noise be renamed? Then it would fall back to default values unless a server owners follows the instructions (would need to be written down) to "port" the old noise they had.
If that's not possible, a setting `mgv7_floatlands_params_are_safe` (defaults to false on existing worlds, true on new worlds) would also avoid this situation.

> Adding water to floatlands is, i now realise, extremely dangerous. I now feel that random floatlands placed above the surface should probably be waterless.

This is less of a problem when moving the floatlands to e.g. y=5000, which would also fit better with the idea that you need a portal to reach them.
Still, fixing water leakage would be nice and I understand that it's hard.

> A complete set of floatland lake and land biomes have to be defined. The standard set of biomes have to be altered to stop just below the floatlands. [...]

I consider this flexibility a feature. There are good reasons why mods may want to define biomes or ores that only appear in floatlands.
In minetest_game, adding all land biomes to floatlands should be doable within ~150 LOC by using a generic registration function that can register biomes/decorations/ores twice with the correct `y_min` and `y_max` parameters.

> The complex biome code means that most games, which do not have it, are incompatible with mgv7 floatlands.

I don't think this is a problem at all. If a game author wants floatlands, they should make the effort to support them and decide exactly how they should look.
If the game author doesn't want/doesn't care about floatlands, the engine deciding to extend land biomes to some floating rocks will rarely make sense.

**Comment by paramat** [comment]
> I don't think floatlands absolutely need to be removed from mgv7

I agree, i could possibly rewrite them to be waterless and created from a new 3D noise, if other core devs insist, although i would rather not. The important thing here is to remove the complex biome code from MTG and avoid floatlands with lakes that require that biome code.
> This is subjective, I think floatlands as they are now look okay even if they use the same "crazy" mountain noise.

Yes, just a subjective personal opinion.
> Can't the noise be renamed?

Yes, probably. However, this is somewhat irrelevant as the important thing is to remove the lakes (and the complex MTG biome code), which requires a complete rewrite or removal.
> I consider this flexibility a feature. There are good reasons why mods may want to define biomes or ores that only appear in floatlands.

I agree, i have no problem with games defining multiple levels of biomes for whatever reasons. You misunderstand what i wrote.
> I don't think this is a problem at all.

It is significant because it is better that games can be compatible with floatlands with no alteration, instead of adding a lot of complexity just to be compatible with the floatlands of 1 mapgen that have several problems, and the author considers a mistake. Any other floatlands i code will be different and not require this game code. So it causes a lot of headache for 1 mistake.
>  the engine deciding to extend land biomes to some floating rocks will rarely make sense.

This works fine and it always makes sense, as the floatlands are just treated like the tops of high mountains.
With the current mgv7 floatlands, a game with simple land biomes results in a horrible mess with underwater surace nodes and tree trunks.

///////////////////////////////////

So, anyway ... if other core devs (2 or 3) insist, i can rewrite mgv7 floatlands to a completely new implementation, with no lakes and spread throughout the atmosphere, using a new 3D noise. However, i prefer to just remove them and not mess with such an old, and now default, mapgen.
The complex biome code would still be removed from MTG, this is the most important part of these changes that i will not compromise on.

I will wait another 2 weeks or so for further input from other core devs, if none i will go ahead with my original plan.
The mgv7 floatlands are half-finished, untuned, have always been disabled by default and described as 'highly unstable', a set of biomes for them has never been added to MTG.
These have been a rather personal experimental project of mine.
So, as long as i do not cause severe problems for users (which my plan does not) i think i have some authority over the concept and future of them.

**Comment by paramat** [comment]
FaceDeer,
> Are there any features of the Floatlands that would be difficult to extract into a Lua mapgen mod? Having the Floatlands as a separate mod

I consider the mgv7 floatlands, as they currently are, something that should have been a Lua mod due to the problematic implementation. They can be done in a Lua mapgen mod yes.
However, a Lua mapgen is problematic for a server due to the long server lags caused during generation. Which is why an engine implementation is very helpful.

**Comment by v-rob** [comment]
Well, for input from a player, I've always found floatlands worthless.  Sure, the mountains are extremely crazy and kinda cool in places, but I would never build on them.  Honestly, when you are on a floatland, it just feels like a big continent, so why not just play on the surface?  Floatlands would be much better if they were more like floating islands somewhere in the ballpark of 100-500 nodes diameter at most.  Otherwise, they're just continents with little advantage above the surface.

**Comment by paramat** [comment]
Just to notify in advance:
Because the mgv7 floatlands are an unfinished personal experiment, that i therefore have authority over, once i start work on the PR (and the relevant PR in MTG) i will then merge these on my own approval. I know the code extremely well (because i wrote it) and it is the sort of mapgen stuff 'i can do in my sleep'.

v-rob, interesting point about them being like continents. My new approach is more like you describe, with more separated and smaller floatlands.

**Comment by sfan5** [comment]
> once i start work on the PR (and the relevant PR in MTG) i will then merge these on my own approval

I understand your reasoning but I will say it again: This is a obvious conflict to the push access [guidelines](https://dev.minetest.net/Git_Guidelines) that were set for this repository, which you have agreed to when becoming a coredev.

**Comment by hecktest** [comment]
Leave floatlands alone! I have plans for them.
Or, give me a better emerge hook than on_generated so that I can delegate the mapgen to some ffi DLL and not have to rely on core mapgens. (anything goes as long as I can multithread it)
![floats](https://user-images.githubusercontent.com/42101236/69060071-04d7d280-0a17-11ea-812e-879fa807971a.jpg)

**Comment by paramat** [comment]
sfan5, sorry for my previous post, i worded that badly, i edited it.
If i work on this, once the PR is complete i will wait 1 week to give other core devs a chance to review, obviously i would prefer to gain a full approval, or at least a code approval.

As explained above, in this particular situation i think i have authority over the future of this feature, and can approve the concept myself.
Because this is code i wrote and understand far more than anyone, and because it will be removal of a feature, i think i can be trusted with the code and can merge this even if it has no code approvals after 1 week, if it comes to that.
The development rules are not rigid rules, in special situations we can and should be flexible. This is why i am giving lots of warning beforehand so we can discuss being flexible here.

I have left this 2 weeks, this has given me a useful chance to think it over, and i still have the same opinons as expressed earlier (sorry hecktest, however, see my offered compromise below).
If this task is not done, i will never finish this feature, it will remain a half finished, disabled and 'highly unstable' mistake in mgv7, making future work on the mapgen more problematic and causing more maintenance and more bugs.
Obviously i, and we, cannot allow that to happen. So there is no benefit at all in leaving this present.

However, as discussed earlier: If 2+ core devs insist (i assume you are 1 of these), i can, after removal, re-add a similar-looking floatland realm to mgv7 but done the right way: no lakes, and using a new dedicated 3D noise which can then be customisable and independent from the mountain noise, and therefore can be tuned to look better. This way i will be happy with the mgv7 floatlands.

The really important part of this change is the change to MTGame, removing that colossal mistake of complex specialist biome code just for one mapgen's floatlands. This is why the lakes have to be removed, the floatland biomes will simply be the high altitude land biomes, just like the tops of mountains. So this way, only the lakes are lost from the feature.

What do core devs think about this? This is my offered compromise, let me know.
I will wait at least another week before starting work.

**Comment by paramat** [comment]
Ok, replacing the mgv7 floatlands seems the right thing to do, so i will do that.
The re-add PR will be after the removal PR, it will be too complex to develop and replace in 1 PR.
I intend to have the replacement ready for 5.2.0.

The new implementation will be better, with a dedicated 3D noise for the floatlands that will therefore be freely customisable. They will have no lakes, but this makes it possible to have multiple levels throughout the atmosphere. I could add parameters for ymin, ymax and level spacing. This way the whole world can be used if desired.

I might start work after this weekend (7-8 December). MTG will have the mgv7 floatland biome code removed as it will no longer be necessary.
Once the PRs are ready i will wait a week for reviews/approval, but if none i will then merge for reasons explained earlier.
